### PR TITLE
Retyped retrieveRecords

### PR DESCRIPTION
**Pull request by Banshaan**
retrieveRecords function makes use of the getMetadataReports, getUserHelp and loadMetadataCount functions that make use of back-end functions.

### CODE REVIEW

File: src/app/shared/interfaces/debug-variables.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+  status?: string;
+  storageId?: string;
+  storageSize?: string;
+  component?: string;
+  conversationId?: string;
+  endpointName?: string;
+  timestamp?: string;
+
+  [key: string]: string | undefined;
+}
```
Comment (philipsens) [inline src/app/shared/interfaces/debug-variables.ts:17]: Can everything be undefined? :o

Comment (Matthbo) [reply src/app/shared/interfaces/debug-variables.ts:17]: This looks like optional variables meant for debugging.
I guess you could set the default for all of them to false unless a different value is given but this is also fine imo

**Review by Matthbo — DISMISSED** [review]

File: src/app/shared/interfaces/debug-variables.ts (hunk 1 above)
Comment (Banshaan) [reply src/app/shared/interfaces/debug-variables.ts]: I checked where these variables are used and none of them need to be undefined, so i changed it.

Comment (Matthbo) [reply src/app/shared/interfaces/debug-variables.ts]: Does the API always give those values back?

Comment (philipsens) [reply src/app/shared/interfaces/debug-variables.ts]: Hm yea, I'm also wondering why there is also a variable key. It might be better to have an `extras` value that includes this 'wildcard'.

Comment (Banshaan) [reply src/app/shared/interfaces/debug-variables.ts]: the [key: string]: string is because the html loops over the variable names to populate the cells

Comment (philipsens) [reply src/app/shared/interfaces/debug-variables.ts]: If you're going to loop over it anyway, is there even any point in defining the variables? (normally yes, for type safety, but the wildcard defeats the purpose)

Comment (Matthbo) [reply src/app/shared/interfaces/debug-variables.ts]: So why can't it be `Record<string,string>` if you're only using it for looping?
And if you do use some variables, split them up into separate types; one with defined properties and the record to loop over

Comment (Banshaan) [reply src/app/shared/interfaces/debug-variables.ts]: im not sure if i understand what you mean

Comment (Banshaan) [reply src/app/shared/interfaces/debug-variables.ts]: Sorry, i didnt see your comment Matthbo. I can it to be `Record<string, string>`, i dont think that'd be an issue.

Comment (Banshaan) [reply src/app/shared/interfaces/debug-variables.ts]: I have changed it to be `Record<string, string>` now

**Review by philipsens — DISMISSED** [review]

**Review by Matthbo — APPROVED** [review]

File: src/app/debug/table/table.component.ts [hunk 2]
```diff
@@ -177,40 +178,54 @@ export class TableComponent implements OnInit, OnDestroy {
...
-          this.toastService.showSuccess('Data loaded!');
-          this.doneRetrieving = true;
-          httpServiceSubscription.unsubscribe();
-        },
-        error: () => {
-          catchError(this.httpService.handleError());
-        },
-      });
+    this.doneRetrieving = false;
+    this.tableSettings.reportMetadata = [];
```
[the remark this answers is not in the document]
Comment (Banshaan) [reply src/app/debug/table/table.component.ts]: Duplicate lines have been deleted

File: src/app/debug/table/table.component.ts [hunk 3]
```diff
@@ -177,40 +178,54 @@ export class TableComponent implements OnInit, OnDestroy {
...
+          next: (debugVariablesList: Record<string, string>[]) => {
+            this.setUniqueOptions(debugVariablesList);
+
+            for (const debugVariables of debugVariablesList) {
+              const listItem: DebugListItem = {
+                checked: false,
+                debugVariables: debugVariables,
+              };
+              this.tableSettings.reportMetadata.push(listItem);
+            }
```
[the remark this answers is not in the document]
Comment (Banshaan) [reply src/app/debug/table/table.component.ts]: I moved this to a separate method

File: src/app/debug/table/table.component.ts [hunk 4]
```diff
@@ -177,40 +178,54 @@ export class TableComponent implements OnInit, OnDestroy {
...
-          this.doneRetrieving = true;
-          httpServiceSubscription.unsubscribe();
-        },
-        error: () => {
-          catchError(this.httpService.handleError());
-        },
-      });
+    this.doneRetrieving = false;
+    this.tableSettings.reportMetadata = [];
+    if (this.viewSettings.currentView?.metadataNames && this.viewSettings.currentView?.storageName) {
```
[the remark this answers is not in the document]
Comment (Banshaan) [reply src/app/debug/table/table.component.ts]: Turned the assertion into a method.

File: src/app/debug/table/table.component.ts [hunk 5]
```diff
@@ -177,40 +178,54 @@ export class TableComponent implements OnInit, OnDestroy {
...
 
   getUserHelp(): void {
-    this.httpService
-      .getUserHelp(this.viewSettings.currentView.storageName, this.viewSettings.currentView.metadataNames)
-      .subscribe({
-        next: (response) => {
-          this.tableSettings.metadataHeaders = response;
-        },
-      });
+    if (this.viewSettings.currentView?.storageName && this.viewSettings.currentView?.metadataNames) {
```
[the remark this answers is not in the document]
Comment (Banshaan) [reply src/app/debug/table/table.component.ts]: Turned the assertion into a method.

File: src/app/debug/table/table.component.ts [hunk 6]
```diff
@@ -175,12 +175,14 @@ export class TableComponent implements OnInit, OnDestroy {
     this.showMultipleFilesSubscription.unsubscribe();
   }
 
+  doStorageAndMetadataNamesExist(): boolean {
+    return this.viewSettings.currentView?.metadataNames && this.viewSettings.currentView?.storageName ? true : false;
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:179]:
```suggestion
    return !!this.viewSettings.currentView?.metadataNames && !!this.viewSettings.currentView?.storageName;
```

File: src/app/test/test-folder-tree/test-folder-tree.component.ts [hunk 7]
```diff
@@ -20,7 +20,7 @@ export class TestFolderTreeComponent {
     folderBehaviourOnClick: 'select',
     expandAllFolders: true,
     highlightOpenFolders: false,
-    autoSelectCondition: (item: CreateTreeItem) => this.autoSelectItem(item),
+    autoOpenCondition: (item: CreateTreeItem) => this.autoSelectItem(item),
```
Comment (MatthijsSmets) [inline src/app/test/test-folder-tree/test-folder-tree.component.ts:23]: Why did you change this?

Comment (MatthijsSmets) [inline src/app/test/test-folder-tree/test-folder-tree.component.ts:23]: Why did you change this?

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]

File: src/app/test/test-folder-tree/test-folder-tree.component.ts (hunk 7 above)
Comment (Banshaan) [reply src/app/test/test-folder-tree/test-folder-tree.component.ts:23]: It told me that it was wrong and it caused a compile error

Comment (Banshaan) [reply src/app/test/test-folder-tree/test-folder-tree.component.ts:23]: with "wrong" i mean it said it didnt exist on FileTreeOptions

Comment (MatthijsSmets) [reply src/app/test/test-folder-tree/test-folder-tree.component.ts:23]: Did you run `yarn install`?

Comment (Banshaan) [reply src/app/test/test-folder-tree/test-folder-tree.component.ts:23]: i did not, i will do so immediately

Comment (Banshaan) [reply src/app/test/test-folder-tree/test-folder-tree.component.ts]: i ran `yarn install` and changed it back

File: src/app/debug/table/table.component.html [hunk 8]
```diff
@@ -96,8 +96,13 @@ <h4 class="progress-report-timout">[One or more reports are in progress for more
           <td class="table-row-checkbox" [style.padding]="getTableSpacing()">
             <input type="checkbox" [checked]="row.checked" (click)="toggleCheck(row)" [style.width]="getCheckBoxSize()" [style.height]="getCheckBoxSize()">
           </td>
-          <td *ngFor="let metadataName of viewSettings.currentView?.metadataNames" [style.padding]="getTableSpacing()" class="table-cell" (click)="openReport(row.debugVariables.storageId)" [title]="row.debugVariables[metadataName]">
-            {{ row.debugVariables[metadataName] | tableCellShortener }}
+          <td 
+          *ngFor="let metadataName of viewSettings.currentView?.metadataNames" 
+          [style.padding]="getTableSpacing()" 
+          class="table-cell" 
+          (click)="openReport(row.debugVariables.storageId)"
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.html:103]: I think it's better to place this (click) on the table row instead of each cell

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]

**Review by MatthijsSmets — APPROVED** [review]

File: src/app/shared/services/http.service.ts [hunk 9]
```diff
@@ -50,8 +50,8 @@ export class HttpService {
     filterHeader: string[],
     metadataNames: string[],
     storage: string,
-  ): Observable<Report[]> {
-    return this.http.get<Report[]>('api/metadata/' + storage + '/', {
+  ): Observable<Record<string, string>[]> {
+    return this.http.get<Record<string, string>[]>('api/metadata/' + storage + '/', {
```
Comment (philipsens) [inline src/app/shared/services/http.service.ts:54]:
```suggestion
    return this.http.get<Record<string, string>[]>(`api/metadata/${storage}/`, {
```

File: src/app/debug/table/table.component.ts [hunk 10]
```diff
@@ -612,19 +649,23 @@ export class TableComponent implements OnInit, OnDestroy {
...
-      if (key.length > longestViewName.length) {
-        longestViewName = key;
+    if (this.viewSettings.views) {
+      let longestViewName = '';
+      for (let key of Object.keys(this.viewSettings.views)) {
+        if (key.length > longestViewName.length) {
+          longestViewName = key;
+        }
       }
+      this.viewDropdownBoxWidth = longestViewName.length / 2 + 'rem';
```
Comment (philipsens) [inline src/app/debug/table/table.component.ts:667]:
```suggestion
      this.viewDropdownBoxWidth = `{longestViewName.length / 2}rem`;
```

File: src/app/debug/table/table.component.ts [hunk 11]
```diff
@@ -521,7 +558,7 @@ export class TableComponent implements OnInit, OnDestroy {
       .reduce((totalQuery: string, selectedReport: any) => totalQuery + 'id=' + selectedReport.storageId + '&', '');
```
Comment (philipsens) [inline src/app/debug/table/table.component.ts:558]:
```suggestion
      .reduce((totalQuery: string, selectedReport: any) => `${totalQuery}id=${selectedReport.storageId}&`, '');
```

File: src/app/debug/table/table.component.ts [hunk 12]
```diff
@@ -181,42 +179,56 @@ export class TableComponent implements OnInit, OnDestroy {
...
+    if (this.viewSettings.currentView?.metadataNames && this.viewSettings.currentView?.storageName) {
+      this.doneRetrieving = false;
+      this.tableSettings.reportMetadata = [];
+      const httpServiceSubscription = this.httpService
+        .getMetadataReports(
+          this.tableSettings.displayAmount,
+          this.tableSettings.filterValues,
+          this.tableSettings.filterHeaders,
+          this.viewSettings.currentView.metadataNames,
+          this.viewSettings.currentView.storageName,
```
Comment (philipsens) [inline src/app/debug/table/table.component.ts:191]: We might add a convention for a Method with more than 2 parameters, but for now it's fine.

**Review by philipsens — CHANGES_REQUESTED** [review]
Everything is fine except some strings: https://github.com/frankframework/frontend-conventions/blob/master/typescript.md#string-manipulation

File: src/app/debug/table/table.component.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (MatthijsSmets) [reply src/app/debug/table/table.component.ts:182]: This line remains unchanged though?

File: src/app/debug/table/table.component.ts [hunk 13]
```diff
@@ -371,22 +395,25 @@ export class TableComponent implements OnInit, OnDestroy {
...
-      .getReport(reportTab.storageId, this.viewSettings.currentView.storageName)
-      .subscribe((report: Report): void => {
-        const reportData: ReportData = {
-          report: report,
-          currentView: this.viewSettings.currentView,
-        };
-        this.tabService.openNewTab(reportData);
-      });
+    const reportTab: DebugListItem = this.tableSettings.reportMetadata.find((report: DebugListItem) => report.checked);
+    const currentView: View = this.viewSettings.currentView ?? ({} as View);
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:399]: Why are you setting view to an empty object?

File: src/app/debug/table/table.component.ts [hunk 14]
```diff
@@ -553,9 +568,13 @@ export class TableComponent implements OnInit, OnDestroy {
...
   downloadReports(exportBinary: boolean, exportXML: boolean): void {
-    const queryString: string = this.tableSettings.reportMetadata
-      .filter((report) => report.checked)
-      .reduce((totalQuery: string, selectedReport: any) => `${totalQuery}id=${selectedReport.storageId}&`, '');
+    const selectedReports: any = this.tableSettings.reportMetadata.filter((report) => report.checked);
+    let queryString = '';
+
+    for (const report of selectedReports) {
+      queryString += `id=${report.storageId}`;
+    }
```
[the remark this answers is not in the document]
Comment (philipsens) [reply src/app/debug/table/table.component.ts:576]: What do you think @MatthijsSmets

[the remark this answers is not in the document]
Comment (philipsens) [reply src/app/debug/table/table.component.ts:577]: By the way, I think you are missing an ampersand.

[the remark this answers is not in the document]
Comment (MatthijsSmets) [reply src/app/debug/table/table.component.ts:577]: > What do you think @MatthijsSmets

I agree, we should take the Big O notation into consideration

File: src/app/debug/table/table.component.ts [hunk 15]
```diff
@@ -350,14 +386,14 @@ export class TableComponent implements OnInit, OnDestroy {
     }
   }
 
-  getStatusColor(metadata: any): string {
-    let statusName = this.viewSettings.currentView.metadataNames.find((name: string) => {
+  getStatusColor(metadata: DebugListItem): string {
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:389]: Right now it's doing an unnecessary find for an attribute that matches the string `status`, this can be removed:
```ts
    getStatusColor(metadata: any): string {
    if (metadata.debugVariables.status) {
      const status = metadata.debugVariables.status;
      if (status.toLowerCase() === 'success') {
        return '#c3e6cb';
      } else if (status.toLowerCase() === 'null') {
        return '#A9A9A9FF';
      } else {
        return '#f79c9c';
      }
    }
    return 'none';
  }
```

File: src/app/shared/interfaces/view-settings.ts
Comment (MatthijsSmets) [inline src/app/shared/interfaces/view-settings.ts:1]: Seeing as view settings only has 2 attributes, it feels redundant, I think it would make more sense to set `currentView` and `views` directly in the component as class variable. That way `viewSettings` also does not have to be initialized with an empty object. The currentView can then be set to the ngModel in the html as well, it might solve your problems with the viewdropdown

File: src/app/debug/table/table.component.ts [hunk 16]
```diff
@@ -568,7 +625,7 @@ export class TableComponent implements OnInit, OnDestroy {
   }
 
   setUniqueOptions(data: any): void {
-    for (const headerName of this.viewSettings.currentView.metadataNames as string[]) {
+    for (const headerName of this.viewSettings.currentView?.metadataNames as string[]) {
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:628]: In the `softFilterList()` method you check if the currentView is not null, why don't you do that here as well.

File: src/app/debug/table/table.component.ts [hunk 17]
```diff
@@ -540,7 +597,7 @@ export class TableComponent implements OnInit, OnDestroy {
       for (let report of data) {
         const reportData: ReportData = {
           report: report,
-          currentView: this.viewSettings.currentView,
+          currentView: this.viewSettings.currentView ?? ({} as View),
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:600]: Setting it to an empty object if it's null should not be the solution here.

File: src/app/debug/table/table.component.ts [hunk 18]
```diff
@@ -478,10 +525,14 @@ export class TableComponent implements OnInit, OnDestroy {
   }
 
   openReport(storageId: string): void {
-    this.httpService.getReport(storageId, this.viewSettings.currentView.storageName).subscribe((data: Report): void => {
-      data.storageName = this.viewSettings.currentView.storageName;
-      this.openReportEvent.next(data);
-    });
+    if (storageId && this.viewSettings.currentView?.storageName) {
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:528]: `storageId` has type string so it can never be null here, so this if statement actually reads: 
```ts
    if (true && this.viewSettings.currentView?.storageName) {
```

File: src/app/debug/table/table.component.ts [hunk 19]
```diff
@@ -418,30 +463,32 @@ export class TableComponent implements OnInit, OnDestroy {
...
+          };
+        },
 
-      complete: () => {
-        this.tabService.openNewCompareTab(compareReports);
-      },
-    });
+        complete: () => {
+          this.tabService.openNewCompareTab(compareReports);
+        },
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:489]: This complete should not be here, if an error occurs in this http request it will still try to open a new compare tab, we've talking about this last friday.

**Review by MatthijsSmets — CHANGES_REQUESTED** [review]

**Review by philipsens — APPROVED** [review]
I have no further comments!

**Comment by MatthijsSmets** [comment]
Also, the linter is not happy

**Comment by Banshaan** [comment]
> Also, the linter is not happy, this should be run before every commit automatically. Did you use --no-verify to override it?

I directly commit from vscode, maybe that's why? I fix any linter issues that are shown while coding though.

File: src/app/debug/table/table.component.ts (hunk 19 above)
Comment (Banshaan) [reply src/app/debug/table/table.component.ts]: Yeah we did, I'm not sure why that's there again but I fixed it.

File: src/app/debug/table/table.component.ts (hunk 18 above)
Comment (Banshaan) [reply src/app/debug/table/table.component.ts]: You're totally right, i got rid of it

File: src/app/debug/table/table.component.ts (hunk 17 above)
Comment (Banshaan) [reply src/app/debug/table/table.component.ts]: Youre totally right, i changed it to check if currentView is defined before  trying to set it to the currentView

File: src/app/debug/table/table.component.ts (hunk 16 above)
Comment (Banshaan) [reply src/app/debug/table/table.component.ts]: I believe i did it because it was easier and less code than an if statement, but an if statement would be cleaner i think so i changed it to an if statement.

File: src/app/shared/interfaces/view-settings.ts
Comment (Banshaan) [reply src/app/shared/interfaces/view-settings.ts:1]: I made currentView and views class variables and deleted the viewSettings interface. It did not, however, fix the problems with the viewdropdown.

File: src/app/debug/table/table.component.ts [hunk 20]
```diff
@@ -225,15 +246,18 @@ export class TableComponent implements OnInit, OnDestroy {
...
-    this.viewSettings.currentViewName = event.target.value;
-    this.clearFilters();
-    this.debugReportService.changeView(this.viewSettings.currentView);
-    this.selectedRow = -1;
-    this.filterService.setMetadataLabels(this.viewSettings.currentView.metadataLabels);
-    this.viewChange.next(this.viewSettings.currentViewName);
+  changeView(): void {
+    if (this.viewSettings.views && this.selectedView) {
+      this.allRowsSelected = false;
+      this.viewSettings.currentView = this.viewSettings.views[this.selectedView.name];
```
[the remark this answers is not in the document]
Comment (Banshaan) [reply src/app/debug/table/table.component.ts]: Yeah you're right, i fixed it

File: src/app/shared/services/http.service.ts [hunk 21]
```diff
@@ -50,8 +50,8 @@ export class HttpService {
     filterHeader: string[],
     metadataNames: string[],
     storage: string,
-  ): Observable<Report[]> {
-    return this.http.get<Report[]>('api/metadata/' + storage + '/', {
+  ): Observable<Record<string, string>[]> {
+    return this.http.get<Record<string, string>[]>(`api/metadata/${storage}/`, {
```
[the remark this answers is not in the document]
Comment (Banshaan) [reply src/app/shared/services/http.service.ts]: It was because there were functions that were used in retrieveRecords that required it to be of type Record<string, string> so it could be looped over. I fixed this so it's still being looped over while being of type Report

[the remark this answers is not in the document]
Comment (MatthijsSmets) [reply src/app/shared/services/http.service.ts]: This function is still Record<string, string> though?

File: src/app/shared/interfaces/view-settings.ts
Comment (MatthijsSmets) [reply src/app/shared/interfaces/view-settings.ts:1]: I still see the ViewSettings class committed here?

File: src/app/debug/table/table.component.ts [hunk 22]
```diff
@@ -562,31 +606,48 @@ export class TableComponent implements OnInit, OnDestroy {
...
-        }
-        if (element[upperHeaderName]) {
-          uniqueValues.add(element[upperHeaderName]);
-        }
-        if (element[headerName]) {
-          uniqueValues.add(element[headerName]);
-        }
+  setUniqueOptions(data: Report[]): void {
+    if (this.currentView) {
+      for (const headerName of this.currentView.metadataNames as string[]) {
```
Comment (MatthijsSmets) [inline src/app/debug/table/table.component.ts:611]:
```suggestion
      for (const headerName of this.currentView.metadataNames) {
```

**Comment by MatthijsSmets** [comment]
On your branch, if I switch from white box view to black box view it does not hide the corresponding checkpoints

**Comment by MatthijsSmets** [comment]
Closing this because it has been open too long and merge conflicts.
